app/models/host/managed.rb

```
@@ -340,21 +340,15 @@ def host_params
     @cached_host_params = hp
   end
 
-  def self.importHostAndFacts yaml
-    facts = YAML::load yaml
-    case facts
-      when Puppet::Node::Facts
-        certname = facts.name
-        name     = facts.values["fqdn"].downcase
-        values   = facts.values
-      when Hash
-        certname = facts["clientcert"] || facts["certname"]
-        name     = facts["fqdn"].downcase
-        values   = facts
-        return raise(::Foreman::Exception.new(N_("invalid facts hash"))) unless name and values
-      else
-        return raise(::Foreman::Exception.new(N_("Invalid Facts, much be a Puppet::Node::Facts or a Hash")))
-    end
+  def self.importHostAndFacts json
+    facts = JSON.parse(json)
+    raise(::Foreman::Exception.new(N_("Invalid Facts, much be a Hash"))) unless facts.is_a?(Hash)
+
+    # we expect a "name" and a "values" part to the uploaded data
+    values   = facts['values']
+    name     = facts['name'].downcase || values["fqdn"].downcase
+    certname = values["clientcert"] || values["certname"]
```

separate to "name"? what I'm working from is what you get in the Puppet yaml cache, which is basically `{name=>'name', values=>{facts...}}`. We can add certname as a firstclass param, but we'll have to manipulate the data more in `node.rb`. That's not necessarily bad, just being clear on implications.

Assuming we want that, then we can raise errors if either value is missing, or use the fallbacks we have today from the values hash. Thoughts?